rwengine/src/engine/Garage.cpp

```diff
@@ -0,0 +1,887 @@
+                 getDistanceToGarage(playerPosition) >= 5.f) ||
+                (playerIsInVehicle &&
+                 getDistanceToGarage(playerPosition) >= 10.f)) {
+                return true;
+            }
+
+            return false;
+        }
+
+        case Type::MissionToOpenAndClose: {
+            // @todo unimplemented
+            return false;
+        }
+
+        case Type::MissionForSpecificCar: {
+            // @todo unimplemented
+            return false;
+        }
+
+        case Type::MissionKeepCarAndRemainClosed: {
+            // @todo unimplemented
+            return false;
+        }
+
+        default: { return false; }
+    }
+
+    return false;
+}
+
+void Garage::doOnOpenEvent() {
+    auto player = engine->getPlayer();
+    auto plyChar = player->getCharacter();
+    auto playerPosition = plyChar->getPosition();
+    auto playerVehicle = plyChar->getCurrentVehicle();
+    bool playerIsInVehicle = playerVehicle != nullptr;
+    RW_UNUSED(playerPosition);
+    RW_UNUSED(playerIsInVehicle);
+
+    switch (type) {
+        case Type::Mission: {
+            break;
+        }
+
+        case Type::BombShop1: {
+            break;
+        }
+
+        case Type::BombShop2: {
+            break;
+        }
+
+        case Type::BombShop3: {
+            break;
+        }
+
+        case Type::Respray: {
+            break;
+        }
+
+        case Type::CollectCars1:
+        case Type::CollectCars2: {
+            break;
+        }
+
+        case Type::MissionForCarToComeOut: {
+            break;
+        }
+
+        case Type::Crusher: {
+            break;
+        }
+
+        case Type::MissionKeepCar: {
+            break;
+        }
+
+        case Type::Hideout1:
+        case Type::Hideout2:
+        case Type::Hideout3: {
+            break;
+        }
+
+        case Type::MissionToOpenAndClose: {
+            break;
+        }
+
+        case Type::MissionForSpecificCar: {
+            break;
+        }
+
+        case Type::MissionKeepCarAndRemainClosed: {
+            break;
+        }
+
+        default: { break; }
+    }
+}
+
+void Garage::doOnCloseEvent() {
+    auto player = engine->getPlayer();
+    auto plyChar = player->getCharacter();
+    auto playerPosition = plyChar->getPosition();
+    auto playerVehicle = plyChar->getCurrentVehicle();
+    bool playerIsInVehicle = playerVehicle != nullptr;
+    RW_UNUSED(playerPosition);
+    RW_UNUSED(playerIsInVehicle);
+
+    switch (type) {
+        case Type::Mission: {
+            player->setInputEnabled(true);
+            break;
+        }
+
+        case Type::BombShop1:
+        case Type::BombShop2:
+        case Type::BombShop3: {
+            // Find out real value
+            garageTimer = engine->getGameTime() + 1.5f;
+
+            break;
+        }
+
+        case Type::Respray: {
+            // Find out real value
+            garageTimer = engine->getGameTime() + 2.f;
+            playerVehicle->setHealth(1000.f);
```

Comment:
Shouldn't we somehow assert/check there's playerVehicle?

Reply:
debug assert will do

Reply:
There doesn't seem to be one? Only `RW_UNUSED`. I think we should get rid of `playerIsInVehicle` in this function, and explicitly check if `playerVehicle` is valid here.

Reply:
Probably it would be used in the future for other garage types.